packages/core/src/external/carequality/pd/process-incoming-pd.ts

```diff
@@ -0,0 +1,135 @@
+
+function constructErrorResponse(
+  payload: PatientDiscoveryRequestIncoming,
+  codingSystem: string,
+  code: string,
+  error: string
+): PatientDiscoveryResponseOutgoing {
+  return {
+    id: payload.id,
+    timestamp: payload.timestamp,
+    responseTimestamp: new Date().toISOString(),
+    patientMatch: null,
+    xcpdHomeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
+    operationOutcome: {
+      resourceType: "OperationOutcome",
+      id: payload.id,
+      issue: [
+        {
+          severity: "error",
+          code: "processing",
+          details: {
+            coding: [{ system: codingSystem, code: code }],
+            text: error,
+          },
+        },
+      ],
+    },
+  };
+}
+
+function constructNoMatchResponse(
+  payload: PatientDiscoveryRequestIncoming
+): PatientDiscoveryResponseOutgoing {
+  return {
+    id: payload.id,
+    timestamp: payload.timestamp,
+    responseTimestamp: new Date().toISOString(),
+    patientMatch: false,
+    xcpdHomeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
+    operationOutcome: {
+      resourceType: "OperationOutcome",
+      id: payload.id,
+      issue: [
+        {
+          severity: "error",
+          code: "processing",
+          details: {
+            text: "no match found",
+          },
+        },
+      ],
+    },
+  };
+}
+
+function constructMatchResponse(
+  payload: PatientDiscoveryRequestIncoming,
+  patient: Pick<Patient, "id" | "data">
+): PatientDiscoveryResponseOutgoing {
+  return {
+    id: payload.id,
+    timestamp: payload.timestamp,
+    responseTimestamp: new Date().toISOString(),
+    patientMatch: true,
+    externalGatewayPatient: {
+      id: patient.id,
+      system: "000", // TBD	- what is this
+    },
+    patientResource: convertPatientToFHIR(patient),
+    xcpdHomeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
+  };
+}
+
+export async function processIncomingRequest(
+  payload: PatientDiscoveryRequestIncoming,
+  // workaround to allow injecting the behavior since this is an isolated function, not a class
+  mpi: MPI
+): Promise<PatientDiscoveryResponseOutgoing> {
+  try {
+    const patient = validateFHIRAndExtractPatient(payload.patientResource);
+    const matchingPatient = await mpi.findMatchingPatient(patient);
+    if (!matchingPatient) {
+      return constructNoMatchResponse(payload);
+    }
+    return constructMatchResponse(payload, patientMPIToPartialPatient(matchingPatient));
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  } catch (error: any) {
+    switch (error.constructor) {
+      case InternalError:
+        return constructErrorResponse(
+          payload,
+          "1.3.6.1.4.1.19376.1.2.27.3",
+          "InternalError",
+          error.message
+        );
+      case PatientAddressRequestedError:
+        return constructErrorResponse(
+          payload,
+          "1.3.6.1.4.1.19376.1.2.27.1",
+          "PatientAddressRequested",
```

Comment:
We could have these defined inside each error class, just need an abstract superclass, so `constructErrorResponse` could grab these from the instance of that superclass' "implementation".